### ISSUE TITLE
onMount is called multiple times when using enableExternalSource with MobX.

### ISSUE THREAD

**Issue by jaekwak02** [solidjs/solid#1850]
### Describe the bug

I found two bugs when using a combination of MobX, enableExternalSource, onMount, and ref. I was able to reproduce them in StackBlitz.

This is the configuration I used for `enableExternalSource` which is exactly the example given on how to use it.
```
  let id = 0;

  enableExternalSource((fn, trigger) => {
    const reaction = new Reaction(`externalSource@${++id}`, trigger);
    return {
      track: (x) => {
        let next;
        reaction.track(() => (next = fn(x)));
        return next;
      },
      dispose: () => {
        reaction.dispose();
      }
    };
  });
```

Here's the rundown of the bugs:
- onMount will run multiple times if accessing MobX observables. It will behave like createEffect, pretty much.
- Components will completely remount if you access MobX observables inside of ref, and then they change.

### Your Example Website or App

https://stackblitz.com/edit/solidjs-templates-rca9k9?file=src%2FApp.tsx

### Steps to Reproduce the Bug or Issue

1. Go to the example.
2. Open the console so you can see the console.logs
3. Click Increment on Case 1 to see the complete remounting bug
4. Click Increment on Case 2 to see the onMount getting called multiple times bug

### Expected behavior

So far these are the only two cases where using MobX with Solid seemingly breaks. 
- I would expect onMount to only ever be called once, regardless of observable access, and also..
- The ref setting function should only ever get called once. (I think?). The component should at least definitely not completely remount when the observable changes.

I really love using MobX with SolidJS, I think it creates the PERFECT framework, if all the edge cases can be rounded out correctly. MobX just solves so many problems for me, and it seems like if it can work correctly with Solid, it will be a truly unified, fully fine grained reactive State and View. Please consider fixing this!

### Platform

- OS: Windows 11 Version 22H2 (Build 22621.1992)
- Browser: Chrome
- Version: 115.0.5790.171 (Official Build) (64-bit) (cohort: Stable)

### Additional context

I do have a workaround: using `untracked()` from MobX does prevent both issues from happening. However, here too, Solid's own `untrack()` method does not work.

**Comment by ryansolid** [comment]
Thanks for reporting. Yeah this makes sense why it is happening as we just tap into the core computation creation code. 

How to solve is more interesting. That code has no idea that a computation can be non-tracking. Hmm.. and we could hit it in userland code too (as you said untrack doesn't work). Hmm.. trying to decide if is better to have computation creation aware, or if it is better to double patch untrack so that calling it runs both. Mostly want the impact of the solution to add no overhead to people not using this feature.

**Comment by ryansolid** [comment]
Ok I've added a second argument to useExternalSource that takes an untrack function. If provided it will be called alongside Solid's and should work for things like `onMount`. I didn't wait for a minor as this is an experimental feature. Try it in 1.8.8 when released.

**Comment by jaekwak02** [comment]
Thanks Ryan!

**Comment by ylc395** [comment]
I can not find `enableExternalSource` in api document. Where can we find it? @jaekwak02

**Comment by chanon** [comment]
I this the correct way to use this with MobX ?
```
// enable mobx
import { Reaction, untracked } from "mobx";
import { enableExternalSource } from "solid-js";

const enableMobXWithSolidJS = () => {
  let id = 0;
  enableExternalSource((fn, trigger) => {
    const reaction = new Reaction(`externalSource@${++id}`, trigger);
    return {
      track: (x) => {
        let next;
        reaction.track(() => (next = fn(x)));
        return next;
      },
      dispose: () => {
        reaction.dispose();
      },
    };
  }, untracked); // <--- added untracked here
};
enableMobXWithSolidJS();
///
```

When I add the untracked parameter, it seems all the MobX reactivity stops working.

**Comment by chanon** [comment]
OK I'm not sure if this is a bug or I'm doing something wrong.
The issue is iterating over an array using `<For>`. For some reason it isn't updating when additional items are added if the `untracked` parameter is added.

This is the first version without `untracked`
https://stackblitz.com/edit/solidjs-templates-ukoek9?file=src%2FApp.tsx

Then after adding `untracked` the `array.join` render updates but the `<For>` doesn't
https://stackblitz.com/edit/solidjs-templates-msdyn1?file=src%2FApp.tsx

**Comment by mdynnl** [comment]
mobx's `.push` changes the `.length` but the original array remains the same. the first version seems to work because `.length` is implicitly and incorrectly tracked internally inside `For` as there isn't any untrack-ing supplied.

the latter with untrack-ing prevents that as intended. now you'll need something that changes to trigger `For` such as spreading `[...arr]` or even with explicitly tracking `.length` alongside the actual array `arr.length, arr`.

**Comment by chanon** [comment]
Thank you.

I don't completely understand why it works for solid (by using `createMutable` on an array) but not for MobX. Tried to look into the `For` and `mapArray` code but it is a bit over my head.

Anyways, I tried various ways to make it work for MobX with `untracked` and spreading `[...arr]` is one. Another is `arr.slice()` which turns out to be slightly faster according my benchmark.

Using MobX with `makeAutoObservable` and `arr.slice()` on an array (tested with 10,000+ array elements) seems to be alot faster than solid js using `createMutable` on an array, so I am currently satisfied with this approach.

**Comment by mdynnl** [comment]
> it works for solid (by using createMutable on an array)

Yeah, that's correct. `For` and `Index` treat `createMutable` and `createStore` specially by subscribing to the array itself with `$TRACK` then is notified when its property is set.

`.length` doesn't work if the length stays the same. Subscribing to all properties is basically `.slice()`. You'd probably have to go similar path as solid stores to be efficient with a dedicated signal for self notification. Not really sure if `mobx` already has something similar though.

**Comment by chanon** [comment]
Looks like hooking MobX `observe` into a custom implementation of `For` could work. I will try it out.
https://mobx.js.org/intercept-and-observe.html#observe

From some google searching, it is funny to see that Ryan ran into similar issue in his previous project
https://github.com/ryansolid/mobx-jsx/issues/4

**Comment by spion** [comment]
If anyone is curious whether this is possible without slice, it is with a slightly changed version of the `For` component:


```typescript
function accessWithLength<T>(val: T): T {
  if ('length' in (val as any)) return (val as any).length, val;
  return val;
}
/**
 * Creates a list elements from a list
 *
 * it receives a map function as its child that receives a list element and an accessor with the index and returns a JSX-Element; if the list is empty, an optional fallback is returned:
 * ```typescript
 * <For each={items} fallback={<div>No items</div>}>
 *   {(item, index) => <div data-index={index()}>{item}</div>}
 * </For>
 * ```
 * If you have a list with fixed indices and changing values, consider using `<Index>` instead.
 *
 * @description https://www.solidjs.com/docs/latest/api#for
 */
export function For<T extends readonly any[], U extends JSX.Element>(props: {
  each: T | undefined | null | false;
  fallback?: JSX.Element;
  children: (item: T[number], index: Accessor<number>) => U;
}) {
  const fallback = 'fallback' in props && { fallback: () => props.fallback };
  return ('_SOLID_DEV_'
    ? createMemo(
        mapArray(
          () => accessWithLength(props.each),
          props.children,
          fallback || undefined
        ),
        undefined,
        { name: 'value' }
      )
    : createMemo(
        mapArray(
          () => accessWithLength(props.each),
          props.children,
          fallback || undefined
        )
      )) as unknown as JSX.Element;
}
```

Not sure if the `_SOLID_DEV_` pragma will work for non-core code, but I kept it just in case.

The idea is to access the length property (if present) in a tracked context, as mapArray itself doesn't do it.

**Comment by spion** [comment]
@ryansolid curious, do you think adding tracked `lenght` access in `mapArray` would make sense?

**Comment by westbrma** [comment]
> @ryansolid curious, do you think adding tracked `length` access in `mapArray` would make sense?

I hope we can consider this since this in order to officially support mobx

**Comment by ryansolid** [comment]
The problem is I believe `length` is insufficient to solve the problem. Sure, `push` and `shift` would work but setting at an index won't I believe. I am not completely sure the best way to handle this admittedly. Solid doesn't have special mutable array types so this hasn't been something I've been working towards. Even in my mobx-jsx library I still handle arrays without using ObservableArray which I realize is pretty unorthodox for MobX. They do have an approach to diff arrays built in but it benched slower than what we were doing. I'm not sure. I've been looking at proxy-based mutation again more recently with Solid (looking at 2.0 reactivity) but I'm still not sure I'd handle it differently.

**Comment by spion** [comment]
> setting at an index won't I believe

The `For` component doesn't have to observe those, does it? All it needs to do is know how many calls of the closure to do. The closures themselves can track any change in the input argument (or any sub-properties it subscribes to)

Or do you mean adding things at the next index instead of using `push`? I would be surprised if someone does that, it can result with sparse arrays if done incorrectly which is a lot of pain.

**Comment by ryansolid** [comment]
Yes the latter. Setting stuff at an index in general. It can result in sparse arrays if people set outside of the bounds, but it is also how you would swap rows with mutation.

**Comment by spion** [comment]
Setting stuff at an index - in the middle or slightly outside of the bounds - both seem to work. Demo: https://stackblitz.com/edit/solidjs-templates-dwvajz?file=src%2FApp.tsx

FWIW, I'm happy to continue using my own mobx-compatible `For` to be honest if you still have concerns.

**Comment by westbrma** [comment]
nice work @spion, it would be great if we can work this into the solid library but if not it would be nice to have an official write up on how to support mobx as part of the ecosystem

**Comment by scott-cornwell** [comment]
@ryansolid Any conclusion to this? I'm still not able to get Mobx working correctly. Adding `untracked` seems to stop all reactivity. I tried the custom `For` component but it gives me errors. Thanks!

**Comment by ryansolid** [comment]
@scott-cornwell  to my knowledge the fixes resulting from this issue have resolved the problems reported. No one else has reported any related issues, so if you have encountered any I suggest reporting a new issue with reproduction.

**Comment by scott-cornwell** [comment]
@ryansolid Ok yeah I am still trying to get it working, I will try to make a repro. This thread is kind of hard to follow though, was there something else I missed besides upgrading mobx and solid and adding `untracked`? Is the custom <For /> supposed to be necessary? This is the top result on Google so it would be nice if there was a clearer resolution here.

**Comment by westbrma** [comment]
@scott-cornwell You do not need the custom For anymore, the native one is just fine. Try testing basic reactivity first. All you need is this: `let id = 0;
enableExternalSource((fn, trigger) => {
	const reaction = new Reaction(`externalSource@${++id}`, trigger);
	return {
		track: x => {
			let next;
			reaction.track(() => (next = fn(x)));
			return next;
		},
		dispose: () => reaction.dispose()
	};
}, untracked);`

**Comment by scott-cornwell** [comment]
@westbrma Thank you! I was able to get a minimal example working with only that, so the problem must be in my code somewhere. Still not sure why as it's fairly complicated, it displays all the state initially but it does not update. Anyhow, this helped me narrow it down a lot!

**Comment by westbrma** [comment]
@scott-cornwell glad to help. Verify you are using makeAutoObservable({...}) or makeAutoObservable(this) for classes. also in classes make sure you properties are initialized with a value or null, I believe simply declaring the property may not be reactive.

**Comment by scott-cornwell** [comment]
@westbrma Thanks, yeah I am. I'm adding the properties dynamically but then calling makeAutoObservable afterward. That strategy seems to work fine in a minimal app. Problem seems to be something else, still hunting! If I remove `untracked` it works but everything remounts constantly.

**Comment by westbrma** [comment]
@scott-cornwell  just a thought but your mobx version is about 2 years old

[](https://www.npmjs.com/package/mobx?activeTab=versions)

**Comment by scott-cornwell** [comment]
@westbrma Thanks, but yeah already tried that, I'm on the latest now 6.13.6. 

It's very strange. Tried greatly simplifying my store and the component but it just won't render anything based on the mobx store. I even tried updating the mobx and a regular signal in the same onClick but even though it updates the signal and renders that it won't render the updated mobx state in the same component. I log out the store and it all looks correct, has the proxy objects and getter/setters, and the getter reads back the right value, it simply doesn't react to it. Can't repro in my minimal example at all but I can't tell what's different in mine. This is going to take a while, but I'll get it eventually!

**Comment by scott-cornwell** [comment]
I finally figured out my issue yesterday. It's really not that hard, I just have a custom build/plugin system that bundles each module separately and Solid doesn't like it when you don't use the same Solid instance/imports or something. Fixed by attaching Solid to window as a global and using esbuild plugin to resolve all solid imports.
